client/service/src/lib.rs

```diff
@@ -373,136 +374,159 @@ impl<TBl, TCl, TSc, TNetStatus, TNet, TTxPool, TOc> Spawn for
 /// Builds a never-ending future that continuously polls the network.
 ///
 /// The `status_sink` contain a list of senders to send a periodic network status to.
-fn build_network_future<
+async fn build_network_future<
 	B: BlockT,
 	C: BlockchainEvents<B>,
 	H: sc_network::ExHashT
 > (
 	role: Role,
 	mut network: sc_network::NetworkWorker<B, H>,
 	client: Arc<C>,
-	status_sinks: Arc<Mutex<status_sinks::StatusSinks<(NetworkStatus<B>, NetworkState)>>>,
-	mut rpc_rx: TracingUnboundedReceiver<sc_rpc::system::Request<B>>,
+	status_sinks: Arc<status_sinks::StatusSinks<(NetworkStatus<B>, NetworkState)>>,
+	rpc_rx: TracingUnboundedReceiver<sc_rpc::system::Request<B>>,
 	should_have_peers: bool,
 	announce_imported_blocks: bool,
-) -> impl Future<Output = ()> {
+) {
 	let mut imported_blocks_stream = client.import_notification_stream().fuse();
 	let mut finality_notification_stream = client.finality_notification_stream().fuse();
 
-	futures::future::poll_fn(move |cx| {
+	// In the case where the channel containing RPC requests shuts down, we want the network
+	// future to continue running. To make that easier, we make sure that `rpc_rx` never finishes.
+	let mut rpc_rx = rpc_rx.chain(stream::pending()).fuse();
+
+	loop {
 		let before_polling = Instant::now();
 
-		// We poll `imported_blocks_stream`.
-		while let Poll::Ready(Some(notification)) = Pin::new(&mut imported_blocks_stream).poll_next(cx) {
-			if announce_imported_blocks {
-				network.service().announce_block(notification.hash, Vec::new());
+		// This future does the same as `finality_notification_stream.next()`, except that if
+		// multiple events are ready on the stream, only the last one is returned.
+		let mut last_finality_notification = futures::future::poll_fn(|cx| {
```

Comment:
Wouldn't it be possible to initiate this future outside of the `loop` as a `Stream`? Instead of initiating it on each iteration it would be initiated once.